packages/firestore/src/model/document.ts

```diff
@@ -24,55 +24,162 @@ import { FieldPath } from './path';
+ * don't exist there are different reasons why they don't exist and we treat
+ * these all differently in the cache.
  */
-export abstract class MaybeDocument {
-  constructor(readonly key: DocumentKey, readonly version: SnapshotVersion) {}
+export enum DocumentType {
+  /**
+   * A document that is unknown to the client. It doesn't exist locally but
+   * may or may not exist on the server.
+   */
+  UNKNOWN,
 
-  static compareByKey(d1: MaybeDocument, d2: MaybeDocument): number {
-    return DocumentKey.comparator(d1.key, d2.key);
-  }
+  /**
+   * A special case of UNKNOWN document whose contents are unknown but that is
+   * known to exist on the server. This kind of document comes about when the
+   * server accepts an update (i.e. a mutation with a precondition that the
+   * document exists) but the client didn't have any base document to which
+   * the mutation applied.
+   *
+   * Note that even though the server considers the document to exist, the
+   * client still presents a local view as if it didn't exist. This type exists
+   * to allow the client to record the version at which the server said the
+   * document did exist, so that we can avoid flickering back to old versions
+   * while listening for updates.
+   */
+  CONTENTS_UNKNOWN,
 
   /**
-   * Whether this document had a local mutation applied that has not yet been
-   * acknowledged by Watch.
+   * A document that's definitely known not to exist, either because the server
+   * said it was missing or because a local mutation deleted it.
+   */
+  MISSING,
+
+  /**
+   * A document that's known to exist and we know what's inside it.
    */
-  abstract get hasPendingWrites(): boolean;
+  EXISTS
+}
+
+enum DocumentState {
+  /** No mutations applied. Document was sent to us by Watch. */
+  SYNCED,
 
-  abstract isEqual(other: MaybeDocument | null | undefined): boolean;
+  /**
+   * Local mutations applied via the mutation queue. Document is potentially
+   * inconsistent.
+   */
+  LOCAL_MUTATIONS,
 
-  abstract toString(): string;
+  /**
+   * Mutations applied based on a write acknowledgment. Document is potentially
+   * inconsistent.
+   */
+  COMMITTED_MUTATIONS
 }
 
 /**
  * Represents a document in Firestore with a key, version, data and whether the
  * data has local mutations applied to it.
+ *
+ *
+ * A class representing a deleted document.
+ * Version is set to 0 if we don't point to any specific time, otherwise it
+ * denotes time we know it didn't exist at.
+ *
+ * A class representing an existing document whose data is unknown (e.g. a
+ * document that was updated without a known base document).
  */
-export class Document extends MaybeDocument {
-  readonly hasLocalMutations: boolean;
-  readonly hasCommittedMutations: boolean;
+export class Document {
+  static unknown(key: DocumentKey): Document {
+    return new Document(
+      DocumentType.UNKNOWN,
+      key,
+      SnapshotVersion.MIN,
+      ObjectValue.EMPTY,
```

Comment:
Per my comments elsewhere I think I would prefer if we made .data throw for missing / unknown documents (and optionally added `.dataOrEmptyObject` or something so you can explicitly opt into `ObjectValue.EMPTY`